### ISSUE TITLE
[prefer-readonly-parameter-types] just for Arrays

### ISSUE THREAD

**Issue by screendriver** [typescript-eslint/typescript-eslint#2670]
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/prefer-readonly-parameter-types": 2
  }
}
```

```TS
interface Car {
  wheels: Wheel[];
}

function foo(bar: string[]) {
  //
}
```

**Expected Result**

It would be great if I could configure `prefer-readonly-parameter-types` to **just** check Arrays for `readonly` but nothing else.

**Actual Result**

Can't be configured.

**Additional Info**

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `4.4.1` |
| `@typescript-eslint/parser`        | `4.4.1` |
| `TypeScript`                       | `4.0.3` |
| `ESLint`                           | `7.11.0` |
| `node`                             | `14` |

**Comment by bradzacher** [comment]
could you provide more context?
WHY do you want this?

**Comment by screendriver** [comment]
I always prefer immutability. Especially on things like Arrays. When you are working in a huge codebase where many teams are involved I often see bugs that occur because of the mutation of lists. To prevent that I always prefer `readonly` Arrays, but can't enforce that in a CI pipeline. 

So I wish the CI pipeline would fail if normal arrays are used. I know I could introduce something like [list](https://github.com/funkia/list) but that would introduce a new dependency which I also try to avoid sometimes (it depends on the project).

I guess the main problem would be situations where you have to use normal arrays because of a third party library / type definition or because a use case requires it 🤔

**Comment by screendriver** [comment]
Oh, yeah, that's right. For objects this would be nice as well. The main problem right now is that it's not configurable. So either you can enable it with everything (including readonly parameters and so on) or disable it completely. I wish I could set what I want to check.

**Comment by bradzacher** [comment]
But that's what I'm wondering - why would you only want arrays, or only want objects?
It seems like a weird request to ask that the rule checks that function parameters are readonly if and only if the type of the parameter is an array.

Why _wouldn't_ you want it to check all of your function parameters?

For migration purposes?
Wouldn't it be better to turn it on for a subset of files to incrementally migrate, instead of turning on a subset of features? You'll get a lot more granularity to your incremental migration that way.

**Comment by screendriver** [comment]
For example following will throw a linting error with `prefer-readonly-parameter-types` rule enabled:

```ts
import { IncomingHttpHeaders } from 'http';

export function isHeaderValid(headers: IncomingHttpHeaders) { // Parameter should be a read only type
  return headers['x-something'] === 'foo';
}
```

This is something that I do not control and don't want to create my own type for that. That was the reason why I just want to go with Arrays.

PS: this will not warn me that I'm returning a non `readonly` string array. Is there another rule for that?

```ts
export function foo(): string[] {
  return [];
}
```

**Comment by bradzacher** [comment]
> this will not warn me that I'm returning a non readonly string array. Is there another rule for that

Nope.
Returning readonly arrays is not as clear-cut as accepting readonly parameters.
There are many valid use-cases for mutable return values - so a rule to enforce readonly array returns would be cumbersome to use and would false-positive a lot.

> For example following will throw a linting error with prefer-readonly-parameter-types rule enabled

As it should. You're passing a mutable object into a function.
You can usually fix this by wrapping the type in `Readonly<>`:
```ts
export function isHeaderValid(headers: Readonly<IncomingHttpHeaders>) {
  return headers['x-something'] === 'foo';
}
```

This is not unlike what you'd need to do if the input type was an array:
```
// my-module
export type SomeArrayType = string[];

// ./file.ts
import { SomeArrayType } from 'my-module';

export function isThingValid(thing: SomeArrayType) {} // error - array must be readonly

// fix is:
export function isThingValid(thing: Readonly<SomeArrayType>) {}
```

**Comment by screendriver** [comment]
Hmm ok. I guess we can close this issue then 🙂
